packages/gensx/src/checkpoint.ts

```diff
@@ -0,0 +1,170 @@
+// Cross-platform UUID generation
+async function generateUUID(): Promise<string> {
+  try {
+    // Try Node.js crypto first
+    const crypto = await import("node:crypto");
+    return crypto.randomUUID();
+  } catch {
+    // Fallback to browser crypto
+    if (typeof globalThis !== "undefined") {
+      return globalThis.crypto.randomUUID();
+    }
+    // Simple fallback for environments without crypto
+    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, c => {
+      const r = (Math.random() * 16) | 0;
+      const v = c === "x" ? r : (r & 0x3) | 0x8;
+      return v.toString(16);
+    });
+  }
+}
+
+export interface ExecutionNode {
+  id: string;
+  componentName: string;
+  parentId?: string;
+  startTime: number;
+  endTime?: number;
+  props: Record<string, unknown>;
+  output?: unknown;
+  children: ExecutionNode[];
+  metadata?: {
+    logs?: string[];
+    tokenCounts?: {
+      input: number;
+      output: number;
+    };
```

Comment:
These aren't super meaningful without also knowing the provider/model, so you can calculate cost.

```suggestion
    tokenCounts?: {
      input: number;
      output: number;
      provider: string;
      model: string
    };
```

Reply:
We aren't using these fields right now. I expect we'll update the data model when we do add support for token tracking. Will leave until that point if that is ok with you - otherwise I can just remove this stuff for now.

Reply:
Perhaps removing would be the right choice then 🤷